src/components/views/elements/MemberEventListSummary.js

```diff
@@ -0,0 +1,219 @@
+        threshold: React.PropTypes.number,
+        // The function to render events if they are not being summarised
+        renderEvents: React.PropTypes.function,
+        previousEvent: React.PropTypes.object,
+    },
+
+    getInitialState: function() {
+        return {
+            expanded: false,
+        };
+    },
+
+    getDefaultProps: function() {
+        return {
+            summaryLength: 3,
+            threshold: 3,
+            avatarsMaxLength: 5
+        };
+    },
+
+    toggleSummary: function() {
+        this.setState({
+            expanded: !this.state.expanded,
+        });
+    },
+
+    getEventSenderName: function(ev) {
+        if (!ev) {
+            return 'undefined';
+        }
+        return ev.sender.name || ev.event.content.displayname || ev.getSender();
+    },
+
+    renderNameList: function(events) {
+        if (events.length === 0) {
+            return null;
+        }
+        let originalNumber = events.length;
+        events = events.slice(0, this.props.summaryLength);
+        let lastEvent = events.pop();
+
+        let names = events.map((ev) => {
+            return this.getEventSenderName(ev);
+        }).join(', ');
+
+        if (names.length === 0) {
+            return this.getEventSenderName(lastEvent);
+        }
+
+        // Special case the last name. ' and ' might be included later
+        // So you have two cases:
+        if (originalNumber <= this.props.summaryLength) {
+            // name1, name2 and name3
+            names += ' and ';
+        } else {
+            // name1, name2, name3 [and 100 others]
+            names += ', ';
+        }
+        return names + this.getEventSenderName(lastEvent);
+    },
+
+    renderSummary: function(joinEvents, leaveEvents) {
+        let joiners = this.renderNameList(joinEvents);
+        let remainingJoiners = joinEvents.length - this.props.summaryLength;
+        let leavers = this.renderNameList(leaveEvents);
+        let remainingLeavers = leaveEvents.length - this.props.summaryLength;
+        let joinSummary = null;
+        if (joiners) {
+            joinSummary = (
+                <span>
+                    {joiners} {remainingJoiners > 0 ? 'and ' + remainingJoiners + ' others ':''}joined the room
+                </span>
+            );
+        }
+        let leaveSummary = '';
+        if (leavers) {
+            leaveSummary = (
+                <span>
+                    {leavers} {remainingLeavers > 0 ? 'and ' + remainingLeavers + ' others ':''}left the room
+                </span>
+            );
+        }
+        return (
+            <span>
+                {joinSummary}{joinSummary && leaveSummary?'; ':''}
+                {leaveSummary}
+            </span>
+        );
+    },
+
+    renderAvatars: function(events) {
+        let avatars = events.slice(0, this.props.avatarsMaxLength).map((e) => {
+            let onClickAvatar = () => {
```

Comment:
I'd be sorely tempted to push this down into MemberAvatar, maybe with a `viewUserOnClick=true` property, as a way to factor out the common code with MessagePanel.